flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushFilterInCalcIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,139 @@
+                operand(Calc.class, operand(FlinkLogicalTableSourceScan.class, none())),
+                "PushFilterInCalcIntoTableSourceScanRule");
+    }
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        super.matches(call);
+
+        Calc calc = call.rel(0);
+        RexProgram originProgram = calc.getProgram();
+
+        if (originProgram.getCondition() == null) {
+            return false;
+        }
+
+        FlinkLogicalTableSourceScan scan = call.rel(1);
+        TableSourceTable tableSourceTable = scan.getTable().unwrap(TableSourceTable.class);
+        // we can not push filter twice
+        return canPushdownFilter(tableSourceTable);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        Calc calc = call.rel(0);
+        FlinkLogicalTableSourceScan scan = call.rel(1);
+        TableSourceTable table = scan.getTable().unwrap(TableSourceTable.class);
+        pushFilterIntoScan(call, calc, scan, table);
+    }
+
+    private void pushFilterIntoScan(
+            RelOptRuleCall call,
+            Calc calc,
+            FlinkLogicalTableSourceScan scan,
+            FlinkPreparingTableBase relOptTable) {
+
+        RexProgram originProgram = calc.getProgram();
+
+        RelBuilder relBuilder = call.builder();
+        Tuple2<RexNode[], RexNode[]> extractedPredicates =
+                extractPredicates(
+                        originProgram.getInputRowType().getFieldNames().toArray(new String[0]),
+                        originProgram.expandLocalRef(originProgram.getCondition()),
+                        scan,
+                        relBuilder.getRexBuilder());
+
+        RexNode[] convertiblePredicates = extractedPredicates._1;
+        if (convertiblePredicates.length == 0) {
+            // no condition can be translated to expression
+            return;
+        }
+
+        Tuple2<SupportsFilterPushDown.Result, TableSourceTable> pushdownResultWithScan =
+                createTableScanAfterPushdown(
+                        convertiblePredicates,
+                        relOptTable.unwrap(TableSourceTable.class),
+                        scan,
+                        relBuilder);
+
+        SupportsFilterPushDown.Result result = pushdownResultWithScan._1;
+        TableSourceTable tableSourceTable = pushdownResultWithScan._2;
+
+        FlinkLogicalTableSourceScan newScan =
+                FlinkLogicalTableSourceScan.create(scan.getCluster(), tableSourceTable);
+
+        // build new calc program
+        RexProgramBuilder programBuilder =
+                new RexProgramBuilder(newScan.getRowType(), call.builder().getRexBuilder());
+
+        if (!result.getRemainingFilters().isEmpty()) {
+            RexNode[] unconvertedPredicates = extractedPredicates._2;
+            RexNode remainingCondition =
+                    getRemainingConditions(relBuilder, result, unconvertedPredicates);
+
+            programBuilder.addCondition(remainingCondition);
+        }
```

Comment:
result.getRemainingFilters().isEmpty() && extractedPredicates._2.length == 0

Reply:
Not sure what you're referring to here. If no predicates are remaining we don't need a condition?

Reply:
The condition is wrong here. `RexNodeExtractor.extractConjunctiveConditions` gives us convertable predicates and convertible predicates. If we only take the convertable results into the consideration, we may forget to add the unconvertible predicates into the new calc.